src/v/cluster/topics_frontend.cc

```diff
@@ -827,6 +827,119 @@ ss::future<std::error_code> topics_frontend::force_update_partition_replicas(
+                      allocation_constraints constraints;
+                      constraints.add(distinct_from(current));
+                      return local.reallocate_partition(
+                        tn,
+                        {assignment.id,
+                         static_cast<uint16_t>(current.size()),
+                         std::move(constraints)},
+                        assignment,
+                        get_allocation_domain(tn),
+                        nodes_to_exclude);
+                  });
+                force_move_plan::plan_entry entry{
+                  .ntp = {tn.ns, tn.tp, assignment.id}, .before = current};
+                if (!reallocation) {
+                    vlog(
+                      clusterlog.debug,
+                      "no suitable reallocation found for topic {}, partition "
+                      "{}, error: {}",
+                      tn,
+                      assignment.id,
+                      reallocation.error());
+                    entry.after = remaining;
+                    if (entry.after.size() == 0) {
+                        // no replicas left, pick a random broker_shard.
+                        auto random_node = random_generators::random_choice(
+                          remaining_nodes);
+                        auto random_shard = static_cast<uint32_t>(
+                          random_generators::get_int(
+                            0, (int)random_node.broker.properties().cores - 1));
+                        entry.after.push_back(
+                          {random_node.broker.id(), random_shard});
+                    }
+                } else {
+                    entry.after = std::move(reallocation.value().replicas());
+                }
+                result.plan.push_back(std::move(entry));
+            }
+            co_await ss::maybe_yield();
+        }
+        if (!dry_run) {
+          co_await submit_force_move_plan(result);
+        }
+        co_return result;
+    } catch (const topic_table::concurrent_modification_error& e) {
+        // state changed while generating the plan, force caller to retry;
+        vlog(
+          clusterlog.info,
+          "Topic table state changed when generating force move plan: {}",
+          e.what());
+    }
+    co_return errc::concurrent_modification_error;
+}
+
+ss::future<> topics_frontend::submit_force_move_plan(
+  topics_frontend::force_move_plan& plan) {
+    static constexpr std::chrono::seconds timeout{5s};
+    co_await ss::max_concurrent_for_each(
+      plan.plan, 32, [this](force_move_plan::plan_entry& entry) {
+          return force_update_partition_replicas(
+                   entry.ntp,
+                   entry.after,
+                   model::timeout_clock::now() + timeout)
+            .then([&entry](std::error_code ec) {
+                entry.submission_ec = ec;
+                return ss::now();
+            });
+      });
```

Comment:
I think we need to limit the number of moves scheduled concurrently. We could leverage balancer planner limits if this code is moved there. I would recommend to have all partition reallocation related code in `balancer_backend` or `balancer_planner`.